cli/record/record.go

```diff
@@ -0,0 +1,524 @@
+package record
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/cli/log"
+	"github.com/buildbuddy-io/buildbuddy/cli/login"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/buildbuddy-io/buildbuddy/server/util/uuid"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	bespb "github.com/buildbuddy-io/buildbuddy/proto/build_event_stream"
+	bepb "github.com/buildbuddy-io/buildbuddy/proto/build_events"
+	clpb "github.com/buildbuddy-io/buildbuddy/proto/command_line"
+	pepb "github.com/buildbuddy-io/buildbuddy/proto/publish_build_event"
+)
+
+var (
+	flags = flag.NewFlagSet("record", flag.ContinueOnError)
+
+	besBackend   = flags.String("bes_backend", "remote.buildbuddy.io", "BuildBuddy BES backend target")
+	resultsURL   = flags.String("results_url", "https://app.buildbuddy.io", "BuildBuddy results URL")
+	invocationID = flags.String("invocation_id", "", "Invocation ID to use (auto-generated if not specified)")
+
+	usage = `
+usage: bb record [options] <command> [args...]
+
+Records command output and streams it to BuildBuddy.
+
+Example:
+  $ bb record npm install
+  $ bb record make build
+  $ bb record --invocation_id=my-build ./build.sh
+
+The command will be executed locally and its output will be streamed to
+BuildBuddy, where you can view it in the UI.
+`
+)
+
+func HandleRecord(args []string) (int, error) {
+	// Parse flags until we hit a non-flag argument
+	flags.SetOutput(io.Discard)
+	if err := flags.Parse(args); err != nil {
+		if err == flag.ErrHelp {
+			log.Print(usage)
+			flags.SetOutput(os.Stderr)
+			flags.PrintDefaults()
+			return 1, nil
+		}
+		return -1, err
+	}
+
+	cmdArgs := flags.Args()
+	if len(cmdArgs) == 0 {
+		log.Print("error: must provide a command to execute")
+		log.Print(usage)
+		return 1, nil
+	}
+
+	exitCode, err := record(cmdArgs)
+	if err != nil {
+		log.Printf("Failed to record command: %s", err)
+		return exitCode, err
+	}
+	return exitCode, nil
+}
+
+func record(cmdArgs []string) (int, error) {
+	ctx := context.Background()
+
+	iid := *invocationID
+	if iid == "" {
+		iid = uuid.New()
+	}
+
+	log.Debugf("Starting recording session with invocation ID: %s", iid)
+
+	invocationURL := fmt.Sprintf("%s/invocation/%s", *resultsURL, iid)
+	streamingLog := fmt.Sprintf("\033[32mINFO:\033[0m Streaming results to: \033[4;34m%s\033[0m\n", invocationURL)
```

Comment:
nit: you can use `terminal.Esc` (from the CLI package) to produce these escape codes, e.g. `terminal.Esc(32)` instead of `\033[32m` or `terminal.Esc(4, 34)` instead of `\033[4;34m`. The benefit is that they get disabled automatically if stdout is not a terminal, and it's a bit easier to read.

Reply:
Done